### PR TITLE
Sends Reportback from Phoenix API -> Rogue 

### LINKED ISSUE #7183: Rogue to accept rbs that are sent to Phoenix via API.
_FEATURE OVERVIEW_
-------------------
#### User Story
As a developer, I want Rogue to accept reportbacks that are sent to phoenix via API so that we're collecting all reportbacks.

#### Why This Matters
We want to collect ALL reportbacks (some things, like Gambot and the DoSomething Mobile App use Phoenix API to submit reportbacks).

### PR DESCRIPTION
**Pull request by chloealee**
#### What's this PR do?
- When an app hits Phoenix's `/api/v1/campaigns/[nid]/reportback` endpoint, instead of saving directly to Phoenix, it will send along to Rogue first and then Rogue will send back to Phoenix to be saved.
- If crop values aren't set, don't send to Rogue (both on first send and on cron job).

#### How should this be reviewed?
- Hit the `/api/v1/campaigns/[nid]/reportback` endpoint.
screen shot 2016-11-17 at 3 51 11 pm
- Check the Rogue `reportback_items` table for the reportback item.
- Check the Phoenix `dosomething_reportback_file` table for the reportback item. 
- Check that Phoenix's `dosoemthing_rogue_reportbacks` table has the reportback item with all correct data.

#### Relevant tickets
Fixes #7183 

#### Checklist
- [ ] Documentation added for new features/changed endpoints.
- [ ] Tested on staging.
- [ ] Pinged a PM if this is a larger PR that would benefit from some additional testing love
- [ ] Post a message in #phoenix if this includes something that causes a rebuild!

### CODE REVIEW

**Comment by chloealee** [comment]
Dropping this here before I forget! Before I merge, should I notify anyone about this? I think Aaron is out so not sure who else to mention it to?

File: lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc
Comment (angaither) [inline lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:361]: can you add some comments here to explain why we are only sending the reportback if there isn't a transaction id

File: lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module
Comment (angaither) [inline lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: since we are changing this code in multiple places it might make sense to pull this out into a helper function that prepares the item before sending to rouge, but you can do a follow up PR after merge if you want

File: lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc
Comment (angaither) [inline lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:368]: what happens if there's no rbid?

**Review by angaither — CHANGES_REQUESTED** [review]

File: lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc
Comment (chloealee) [reply lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:361]: yes good call!

File: lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc
Comment (chloealee) [reply lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:368]: would it make sense to do a watchdog error here? This would mean that Rogue's `POST` wouldn't have been successful.

File: lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module
Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: ah yes, great call! Since the two are structured a bit differently (ie `$data['crop_x']` vs `$values['crop_x']` and are set to an array vs object `$values['crop_x']` vs `$task->crop_x`) would it be cleanest to write a helper function with conditionals checking if it's in the `$data` format, set to array, else set to the `$task` object?

Comment (angaither) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: or cast the array as an object when you pass it in so you always are dealing with the same format

File: lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc
Comment (angaither) [reply lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:368]: doesn't `rbid` not exist if it's not already saved to the phoenix DB?

I meant more what's the case where the transaction ID doesn't exist & the reportback id doesn't exist, doesn't that need to save the reportback?

File: lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module
Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: ah i see, thank you!

File: lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc
Comment (chloealee) [reply lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:368]: oh yes I see what you mean now, @angaither ! So if there is no transaction ID, then it sends it to Rogue and Rogue sends back to this function and should now have a transaction id. Since it now has a transaction id, it'll move on to the `else` statement and then save.

File: lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module
Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: @angaither this isn't actually working - I will tag you when it is!

Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: @angaither I'm having trouble wrapping my head around this for some reason, can I do this in a separate PR like you suggested above in order to not hold up this PR from being merged?

Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: actually I think I got something to work!

Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: @angaither I'm playing around with this and think it might make it a little confusing to pull out to a helper function to prepare `$data` to send since here, we'd have to send many different variables to the function (`$northstar_id`, `$user`, `$values`, and `$run`) while in `dosomethign_rogue.cron.inc`, it gets these values from the table/from the `$task` object. I think it makes more sense to keep as is unless I'm not understanding correctly?

Comment (angaither) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: @chloealee the reason I'd advocate for pulling it out is we are doing it in multiple places, and I will also be creating a similar object for the migration script. 
We can pass in an array or an object with all the variables and a common function can prepare it 

Similar to[ `dosomething_northstar_create_user`](https://github.com/DoSomething/phoenix/blob/dev/lib/modules/dosomething/dosomething_northstar/dosomething_northstar.module#L226)

Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: @angaither ah okay makes sense - thank you for the example!

File: lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc
Comment (angaither) [inline lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc]: double commented here

File: lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc
Comment (angaither) [inline lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc]: I'd add (mobile app, or sms reportbacks) here to be super clear in the comment

File: lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module
Comment (angaither) [inline lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: this bracket belongs on the row above in drupal, there's a few other times in this PR it's wrong

File: lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module
Comment (angaither) [inline lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: you should probably only send one user value in, either ns id, or drupal user. I know we are starting to prefer NS id, so it's probably safe to only send that.

File: lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module
Comment (angaither) [inline lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: the run should be a part of the values array like the other vars

File: lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc
Comment (angaither) [inline lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc]: should this section be updated to use the transform function as well?

**Review by angaither — CHANGES_REQUESTED** [review]

File: lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module
Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module]: since there is a small percentage of people that do not have NS Ids, should I do a check before this helper function to send `$northstar_id` if there is one and if not, send the `$user`?

File: lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc
Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc]: I was getting a little caught up on this because in order to send it to `dosomething_rogue_transform_reportback`, we'd be sending in only the `$task` object to fill out `$data` array and seemed to not match up exactly so thought it was cleaner just to keep as is?

Comment (angaither) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc]: hmmm, I think it makes it a little less clear if there's multiple ways this data can be formatted though

Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc]: ohhh okay I get what you are saying now - will change!

Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc]: @angaither sorry - confusion again! Looking at this, I think it's okay to leave as is since it eventually will get to the transformer anyway? The flow for the cron job is that it gets sent to `dosomething_rogue_send_reportback_to_rogue` [here](https://github.com/DoSomething/phoenix/blob/dev/lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc#L49), and in this function, it calls `dosomething_rogue_transform_reportback` so that data will just get formatted eventually anyway. LMK what you think!

Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc]: @angaither i did do some tweaking that should make it better though so I'll push that up as well!

File: lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc
Comment (angaither) [inline lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc]: this line is still commented out?

Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc]: oh right - I will delete that! Just had as a placeholder if we wanted to send to the helper function.

Comment (angaither) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc]: maybe we are just going in circles here, but shouldn't that whole chunk be using the helper function?

Comment (chloealee) [reply lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc]: @angaither omg YES sorry this took me so long to wrap my head around! All fixed :)

File: lib/modules/dosomething/dosomething_reportback/dosomething_reportback.forms.inc
Comment (angaither) [inline lib/modules/dosomething/dosomething_reportback/dosomething_reportback.forms.inc]: is this supposed to be commented out still?

**Comment by chloealee** [comment]
closing with fixes from #7227

**Pull request closed without merging**
